### PR TITLE
Add Docker support

### PR DESCRIPTION
**Pull request by ra0x3**
### Description
- This PR:
  - Allows a user to start the service/node via a config file _or_ command line opts, or using nothing at all
  - Adds Docker support (the config stuff was a prerequisite)
  - Changes the name of the executor binary from `service` to `fuel-executor`

Thus addressing:
- https://github.com/FuelLabs/fuel-indexer/issues/47
- https://github.com/FuelLabs/fuel-indexer/issues/33
  - Respecting the discussion in https://github.com/FuelLabs/fuel-indexer/issues/89
- https://github.com/FuelLabs/fuel-indexer/issues/46
- https://github.com/FuelLabs/fuel-indexer/issues/81

### Notes
- I created `fuel-indexer-lib` as a utility all the other crates can pull from (for logic that is not unique to a single crate)
- If someone feels strongly we shouldn't have a config file, I will remove it

#### Executor
- [ ] Try to start the service a few different ways:
  - [ ] Just by running `cargo run --bin fuel-executor ` 
    - The service should use the specified defaults (using sensible defaults for everything)
  - [ ] Update the `config.yaml` with some non-default options and `cargo run --bin fuel-executor -- -c config.yaml`
    - Service should run using these updated options
    - Try adding secrets from the env like `$POSTGRES_PASSWORD` in the config.yaml and these should work as well
  - [ ] Run the service by just passing in CLI opts `cargo r --bin fuel-executor -- --fuel-node-host 127.0.0.1 --fuel-node-port 4004 --graphql-api-host 0.0.0.0 --graphql-api-port 23232 --postgres-user postgres --postgres-host localhost --postgres-port 5432`
    - Service should run using these CLI opts

### Docker
- [ ] Try building the image `docker build -t local/fuel-indexer:latest -f deployment/Dockerfile .`
  - Open an `ngrok` terminal for a public Fuel-node `ngrok http [fuel-node-port]`
  - Open an `ngrok` for a public Postgres instance `ngrok tcp 5432`
- [ ] Run the Docker image with `docker run $IMAGE --fuel-node-host $FUEL_NGROK_HOST --fuel-node-port $FUEL_NGROK_PORT --graphql-api-host 0.0.0.0 --graphql-api-port 23232 --postgres-user postgres --postgres-host $POSTGRES_NGROK_HOST --postgres-port $POSTGRES_NGROK_PORT`

### CODE REVIEW

**Comment by ra0x3** [comment]
@tjsharp1 
- I ping'd @rfuelsh to review the Docker components (I realize that CI step is failing).
- Pinging you for a initial look at the rest of the changes

File: .env.example [hunk 1]
```diff
@@ -0,0 +1,5 @@
+POSTGRES_USER=
+POSTGRES_DATABASE=
+POSTGRES_PASSWORD=
+POSTGRES_HOST=
+POSTRES_PORT=
```
Comment (rfuelsh) [inline .env.example:5]: @ra0x3 - will values be placed here eventually?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -226,6 +226,71 @@ jobs:
...
+
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      - name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v3
+        with:
+          images: |
+            ghcr.io/fuellabs/fuel-core
```
Comment (rfuelsh) [inline .github/workflows/ci.yml:252]: @ra0x3 - you wanna change this to gchr.io repo for fuel-indexer if you intend to create the image e.g. ghcr.io/fuellabs/fuel-indexert

File: deployment/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,51 @@
+# Stage 1: Build
+FROM lukemathwalker/cargo-chef:latest-rust-1.61 as chef
+
+WORKDIR /build/
+
+# hadolint ignore=DL3008
+RUN apt-get update && \
```
Comment (rfuelsh) [inline deployment/Dockerfile:7]: apt-get update -y

File: .env.example (hunk 1 above)
Comment (ra0x3) [reply .env.example:5]: @rfuelsh Yea the idea is that .env.example will show configuration (primarily for secrets) via a .env file

**Comment by ra0x3** [comment]
@Voxelot  @tjsharp1  Just FYI this is officially ready for review

File: indexer-lib/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    };
+    use anyhow::Result;
+    use serde::Deserialize;
+    use std::fmt::Write;
+    use std::net::SocketAddr;
+    use std::path::PathBuf;
+    use structopt::StructOpt;
+
+    #[derive(Debug, StructOpt)]
+    #[structopt(
```
[the remark this answers is not in the document]
Comment (ra0x3) [reply indexer-lib/src/lib.rs:42]: @Voxelot Interesting. I noticed this was different than what was used for `forc` but didn't think "it's not the same" was a good enough reason to change it. I can make this update

**Comment by Voxelot** [comment]
@ra0x3 

For easier and automated testing, there's two strategies I'd suggest.

near-term goals
a.a) Use docker-compose to stand up an environment for localized testing. Docker-compose will handle the tricky port forwarding and networking details between services and is reproducible in CI or other developer machines. This will make it easy to setup a mock installation using all docker images for fuel-core, indexer, and postgres.
a.b) Setup some very basic "Acceptance Tests" that do the bare minimum to ensure the indexer is behaving correctly when connected with all the other live components (e.g. the main functional usecase like indexing events and making that data available via the API)

longer-term goals
b.a)  Have a k8s environment that all merges to master deploy the latest docker image to. Setup a ci flow to trigger your Acceptance Tests to run against every new deployment and provide feedback / slack alerts if something is amiss. 
b.c) AT + Ephemeral env: Once ephemeral environments are better supported, each PR could have its' own unique CD environment that also run the Acceptance Tests post-deploy.

**Comment by ra0x3** [comment]
@Voxelot 
- docker-compose makes a lot of sense
- I've created https://github.com/FuelLabs/fuel-indexer/issues/126 to track this
- For the sake of not keeping this work hanging on the vine too long -- and since we're not using these images at the moment -- I would ask if we're good to merge this (if it looks OK) and then follow this up with #126  ?
  - Obviously I could also see the argument for "do these in one PR" but my thinking here is that let's get as many quick wins as possible while at the same time not putting the follow up work on the shelf for too long (given everything else that needs to be done)

**Comment by Voxelot** [comment]
Does the api-server also need a separate docker image? Or does the fuel-executor binary include both?

**Comment by ra0x3** [comment]
@Voxelot 
- The `fuel-api-server` can indeed be its own standalone binary, but yea `fuel-executor` includes a GraphQL endpoint as well
- I created https://github.com/FuelLabs/fuel-indexer/issues/128 which ideally I'd like to tackle after https://github.com/FuelLabs/fuel-indexer/pull/127

**Review by Voxelot — DISMISSED** [review]

**Comment by Voxelot** [comment]
There's a lot of changes here besides the dockerfile which I didn't have time to extensively review, but the dockerfile itself looks good.

File: deployment/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+COPY . .
+RUN cargo chef prepare --recipe-path recipe.json
+
+
+FROM chef as builder
+ENV CARGO_NET_GIT_FETCH_WITH_CLI=true
+COPY --from=planner /build/recipe.json recipe.json
+RUN cargo chef cook --release -p fuel-indexer --recipe-path recipe.json
+COPY . .
+RUN cargo build --release -p fuel-executor --all-targets
```
Comment (Voxelot) [inline deployment/Dockerfile:32]:
```suggestion
RUN cargo build --release -p fuel-indexer --all-targets
```

**Pull request merged**